File: ios/RetryClassifier.swift
import Foundation

/// Decides what one attempt's result means: the spec's retry, auth and
/// lifetime table (section 6.1). Pure: no I/O, no session state.
enum RetryClassifier {
  enum Class: Equatable {
    case accepted
    case transient
    case auth
    case terminalHttp
    /// The payload is gone. A retry can never succeed.
    case fileMissing
    /// The payload exists but cannot be read now (iOS before first unlock).
    case fileUnreadable
    case expired
  }

  struct Input {
    var statusCode: Int?
    var body: String?
    var error: NSError?
    var accept: [UploadOutcome.AcceptRule]
    var policy: RetryPolicy
    /// Changes nothing: a chunked definition sets `exempt: []`, so the
    /// terminal row applies through the policy. It is here so a test can pin
    /// the part-404 case.
    var isChunkedPart: Bool
    var fileExists: Bool
    var now: Double
    var expiresAt: Double
  }

  /// A cancellation (NSURLErrorCancelled) never reaches here: the caller
  /// handles it from the task's recorded purpose.
  static func classify(_ i: Input) -> Class {
    if i.error == nil, let code = i.statusCode,
       UploadOutcome.isAccepted(code, body: i.body, accept: i.accept) {
      return .accepted
    }
    if i.now >= i.expiresAt { return .expired }
    if let error = i.error {
      if errorKind(for: error) == "file" { return i.fileExists ? .fileUnreadable : .fileMissing }
      return .transient
    }
    guard let code = i.statusCode else { return .transient }
    switch code {
    case 401, 403: return .auth
    case 408, 429, 500...599: return .transient
    case 400...499: return i.policy.exempt.contains(code) ? .transient : .terminalHttp
    // 1xx and 3xx the session did not follow. The answer will not change.
    default: return .terminalHttp
    }
  }

  /// Jittered exponential backoff. `attempt` is 1-based: 1 gives about base.
  /// min(base * 2^(attempt-1), max), times (1 + jitter * (2r - 1)), >= 0.
  static func backoffMs(attempt: Int, policy: RetryPolicy, random: () -> Double) -> Int {
    let exponent = Double(min(max(attempt - 1, 0), 40))
    let raw = min(policy.baseMs * pow(2, exponent), policy.maxMs)
    let jittered = raw * (1 + policy.jitter * (2 * random() - 1))
    return Int(max(jittered, 0).rounded())
  }

  /// The errorKind taxonomy shared with Android: a missing or unreadable
  /// source file is 'file'; other URL-domain errors are 'network'; anything
  /// else is 'unknown'.
  static func errorKind(for error: NSError) -> String {
    switch (error.domain, error.code) {
    case (NSURLErrorDomain, NSURLErrorFileDoesNotExist),
         (NSURLErrorDomain, NSURLErrorCannotOpenFile),
         (NSURLErrorDomain, NSURLErrorNoPermissionsToReadFile),
         (NSCocoaErrorDomain, NSFileNoSuchFileError),
         (NSCocoaErrorDomain, NSFileReadNoSuchFileError),
         (NSCocoaErrorDomain, NSFileReadNoPermissionError):
      return "file"
    case (NSURLErrorDomain, _):
      return "network"
    default:
      return "unknown"
    }
  }

  static func isCancellation(_ error: NSError?) -> Bool {
    error?.domain == NSURLErrorDomain && error?.code == NSURLErrorCancelled
  }
}

File: android/src/main/java/ai/openspace/backgroundupload/RetryClassifier.kt
package ai.openspace.backgroundupload

import java.io.IOException
import kotlin.math.min
import kotlin.random.Random

/**
 * The retry table (plan section 6.1), as pure functions.
 *
 * | Response or failure                         | Verdict           |
 * | 2xx, or an accept rule matches              | Accepted          |
 * | 401, 403                                    | Auth (park)       |
 * | 408, 429, 5xx                               | Transient         |
 * | other 4xx in `exempt` (default [404])       | Transient         |
 * | other 4xx                                   | Terminal http     |
 * | any other status (1xx, a final 3xx)         | Terminal http     |
 * | IOException, payload file missing           | Terminal file     |
 * | IOException                                 | Transient         |
 * | anything else                               | Terminal unknown  |
 */
object RetryClassifier {

  sealed class Verdict {
    object Accepted : Verdict()
    object Transient : Verdict()
    object Auth : Verdict()
    data class Terminal(val errorKind: String, val message: String) : Verdict()
  }

  data class Policy(val baseMs: Long, val maxMs: Long, val jitter: Double, val exempt: List<Int>)

  /** A wait up to this long happens inside the worker. A longer one releases the worker. */
  const val IN_WORKER_BACKOFF_MAX_MS = 30_000L

  fun policy(defaults: RetryDefaults, override: RetryOverride?): Policy = Policy(
    baseMs = override?.baseMs ?: defaults.baseMs,
    maxMs = override?.maxMs ?: defaults.maxMs,
    jitter = override?.jitter ?: defaults.jitter,
    exempt = override?.exempt ?: defaults.exempt,
  )

  fun classifyResponse(
    code: Int,
    body: String?,
    accept: List<UploadOutcome.AcceptRule>,
    exempt: List<Int>,
  ): Verdict = when {
    UploadOutcome.isAccepted(code, body, accept) -> Verdict.Accepted
    code == 401 || code == 403 -> Verdict.Auth
    code == 408 || code == 429 || code in 500..599 -> Verdict.Transient
    code in 400..499 && code in exempt -> Verdict.Transient
    else -> Verdict.Terminal("http", "HTTP $code")
  }

  /** A CancellationException is never classified; the caller rethrows it first. */
  fun classifyFailure(error: Throwable, fileExists: Boolean): Verdict = when {
    error is IOException && !fileExists ->
      Verdict.Terminal("file", "request body file is missing: ${error.message ?: error.javaClass.simpleName}")
    error is IOException -> Verdict.Transient
    else -> Verdict.Terminal("unknown", error.message ?: error.javaClass.simpleName)
  }

  /**
   * The live attempt's errorKind for a transport failure, from its
   * [classifyFailure] verdict: file, unknown, or network for a transient one.
   */
  fun failureKind(verdict: Verdict): String = (verdict as? Verdict.Terminal)?.errorKind ?: "network"

  fun isExpired(now: Long, expiresAt: Long) = now >= expiresAt

  /**
   * base * 2^(streak-1), capped at maxMs, then spread by ± jitter and capped
   * again. streak 1 is baseMs.
   */
  fun backoffMs(policy: Policy, streak: Int, random: Random = Random.Default): Long {
    val exponent = (streak.coerceAtLeast(1) - 1).coerceAtMost(40)
    val raw = min(policy.baseMs.toDouble() * Math.pow(2.0, exponent.toDouble()), policy.maxMs.toDouble())
    val jitter = policy.jitter.coerceIn(0.0, 1.0)
    val spread = raw * (1.0 + jitter * (2.0 * random.nextDouble() - 1.0))
    return min(spread, policy.maxMs.toDouble()).toLong().coerceAtLeast(0L)
  }

  /** The wake time, never later than expiresAt, so an entry expires on time. */
  fun nextAttemptAt(now: Long, backoffMs: Long, expiresAt: Long): Long = min(now + backoffMs, expiresAt)
}
